evm/src/keccak_sponge/keccak_sponge_stark.rs

```diff
@@ -0,0 +1,441 @@
+    Column::sum(iter::once(&cols.is_full_input_block).chain(&cols.is_final_input_len[i..]))
+}
+
+/// Information about a Keccak sponge operation needed for witness generation.
+#[derive(Debug)]
+pub(crate) struct KeccakSpongeOp {
+    // The address at which inputs are read.
+    pub(crate) context: usize,
+    pub(crate) segment: Segment,
+    pub(crate) virt: usize,
+
+    /// The timestamp at which inputs are read.
+    pub(crate) timestamp: usize,
+
+    /// The length of the input, in bytes.
+    pub(crate) len: usize,
+
+    /// The input that was read.
+    pub(crate) input: Vec<u8>,
+}
+
+#[derive(Copy, Clone, Default)]
+pub(crate) struct KeccakSpongeStark<F, const D: usize> {
+    f: PhantomData<F>,
+}
+
+impl<F: RichField + Extendable<D>, const D: usize> KeccakSpongeStark<F, D> {
+    #[allow(unused)] // TODO: Should be used soon.
+    pub(crate) fn generate_trace(
+        &self,
+        operations: Vec<KeccakSpongeOp>,
+        min_rows: usize,
+    ) -> Vec<PolynomialValues<F>> {
+        let mut timing = TimingTree::new("generate trace", log::Level::Debug);
+
+        // Generate the witness row-wise.
+        let trace_rows = timed!(
+            &mut timing,
+            "generate trace rows",
+            self.generate_trace_rows(operations, min_rows)
+        );
+
+        let trace_polys = timed!(
+            &mut timing,
+            "convert to PolynomialValues",
+            trace_rows_to_poly_values(trace_rows)
+        );
+
+        timing.print();
+        trace_polys
+    }
+
+    fn generate_trace_rows(
+        &self,
+        operations: Vec<KeccakSpongeOp>,
+        min_rows: usize,
+    ) -> Vec<[F; NUM_KECCAK_SPONGE_COLUMNS]> {
+        let mut rows = vec![];
+        for op in operations {
+            rows.extend(self.generate_rows_for_op(op));
+        }
+
+        let num_rows = rows.len().max(min_rows).next_power_of_two();
+        let padding_row = self.generate_padding_row();
+        for _ in rows.len()..num_rows {
+            rows.push(padding_row);
+        }
+        rows
+    }
+
+    fn generate_rows_for_op(&self, op: KeccakSpongeOp) -> Vec<[F; NUM_KECCAK_SPONGE_COLUMNS]> {
+        let mut rows = vec![];
+
+        let mut sponge_state = [0u32; KECCAK_WIDTH_U32S];
+
+        let mut input_blocks = op.input.chunks_exact(KECCAK_RATE_BYTES);
+        let mut already_absorbed_bytes = 0;
+        for block in input_blocks.by_ref() {
+            let row = self.generate_full_input_row(
+                &op,
+                already_absorbed_bytes,
+                sponge_state,
+                block.try_into().unwrap(),
+            );
+
+            // xor block into sponge_state's rate elements.
+            let block_u32s = block
+                .chunks(size_of::<u32>())
+                .map(|chunk| u32::from_le_bytes(chunk.try_into().unwrap()));
+            for (state_u32, block_u32) in sponge_state.iter_mut().zip(block_u32s) {
+                *state_u32 ^= block_u32;
+            }
```

Comment:
Hm I'm confused about this. Don't we already XOR the block into the state on line 303?

Reply:
Oh right, forgot to remove this.